gulpfile.js

```diff
@@ -124,28 +172,197 @@ gulp.task("default", function (callback) {
+            var emulatorReadyAttempts = 0;
+            function androidEmulatorReadyLooper() {
+                ++emulatorReadyAttempts;
+                if (emulatorReadyAttempts > emulatorMaxReadyAttempts)
+                {
+                    console.log("Android emulator is not ready after " + emulatorMaxReadyAttempts + " attempts, abort.");
+                    androidProcess.kill();
+                    return callback(1);
+                }
+                setTimeout(androidEmulatorReady.bind(undefined, androidEmulatorReadyLooper), emulatorReadyCheckDelay);
+            }
+            androidEmulatorReadyLooper();
+        }, true);
+    }
+    // kills the iOS emulator then starts it
+    function killThenStartIOS() {
+        spawnCommand("killall", ["\"" + iOSSimulatorProcessName + "\""], () => {
+            spawnCommand("xcrun", ["instruments", "-w", iOSEmulatorName], () => {
+                var emulatorReadyAttempts = 0;
+                function iOSEmulatorReadyLooper() {
+                    ++emulatorReadyAttempts;
+                    if (emulatorReadyAttempts > emulatorMaxReadyAttempts)
+                    {
+                        console.log("iOS emulator is not ready after " + emulatorMaxReadyAttempts + " attempts, abort.");
+                        return callback(1);
+                    }
+                    setTimeout(iOSEmulatorReady.bind(undefined, iOSEmulatorReadyLooper), emulatorReadyCheckDelay);
+                }
+                iOSEmulatorReady(iOSEmulatorReadyLooper);
+            });
+        }, true);
+    }
+    if (!restartIfRunning) {
+        if (android) {
+            androidEmulatorReady(() => {
+                killThenStartAndroid();
+            });
+        }
+        if (ios) {
+            iOSEmulatorReady(() => {
+                killThenStartIOS();
+            });
+        }
+    } else {
+        if (android) killThenStartAndroid();
+        if (ios) killThenStartIOS();
+    }
+    // This needs to be done so that the task will exit.
+    // The command that creates the Android emulator persists with the life of the emulator and hangs this process unless we force it to quit.
+    gulp.doneCallback = (err) => {
+        process.exit(err ? 1 : 0);
+    }
+}
 
-gulp.task("test-ios-uiwebview", function (callback) {
-    var command = "mocha";
-    var args = ["./bin/test", "--mockserver", "http://127.0.0.1:3000", "--platform", "ios", "--use-wkwebview", "false"];
-    executeCommand(command, args, callback);
-});
+// procedurally generate tasks for every possible testing configuration
+var cleanSuffix = "-clean";
+var fastSuffix = "-fast";
 
-gulp.task("test-ios-wkwebview", function (callback) {
-    var command = "mocha";
-    var args = ["./bin/test", "--mockserver", "http://127.0.0.1:3000", "--platform", "ios", "--use-wkwebview", "true"];
-    executeCommand(command, args, callback);
-});
+// generate tasks for starting emulators
+function generateEmulatorTasks(taskName, android, ios) {
```

Comment:
also i think that having it in the testing process is honestly nicer. if you accidentally started tests without having your emulator ready it automatically starts them. on the other hand, if your emulator is already running it has basically zero impact on your runtime.